lib/src/rules/iterable_contains_unrelated_type.dart

```diff
@@ -0,0 +1,198 @@
+is possible for an Iterable to contain elements of type unrelated to its
+parameter type, but this practice also should be avoided.
+
+**BAD:**
+```
+void someFunction() {
+  var list = <int>[];
+  if (list.contains('1')) print('someFunction'); // LINT
+}
+```
+
+**BAD:**
+```
+void someFunction3() {
+  List<int> list = <int>[];
+  if (list.contains('1')) print('someFunction3'); // LINT
+}
+```
+
+**BAD:**
+```
+void someFunction8() {
+  List<DerivedClass2> list = <DerivedClass2>[];
+  DerivedClass3 instance;
+  if (list.contains(instance)) print('someFunction8'); // LINT
+}
+```
+
+**BAD:**
+```
+abstract class SomeIterable<E> implements Iterable<E> {}
+
+abstract class MyClass implements SomeIterable<int> {
+  bool badMethod(String thing) => this.contains(thing); // LINT
+}
+```
+
+**GOOD:**
+```
+void someFunction10() {
+  var list = [];
+  if (list.contains(1)) print('someFunction10'); // OK
+}
+```
+
+**GOOD:**
+```
+void someFunction1() {
+  var list = <int>[];
+  if (list.contains(1)) print('someFunction1'); // OK
+}
+```
+
+**GOOD:**
+```
+void someFunction4() {
+  List<int> list = <int>[];
+  if (list.contains(1)) print('someFunction4'); // OK
+}
+```
+
+**GOOD:**
+```
+void someFunction5() {
+  List<ClassBase> list = <ClassBase>[];
+  DerivedClass1 instance;
+  if (list.contains(instance)) print('someFunction5'); // OK
+}
+
+abstract class ClassBase {}
+
+class DerivedClass1 extends ClassBase {}
+```
+
+**GOOD:**
+```
+void someFunction6() {
+  List<Mixin> list = <Mixin>[];
+  DerivedClass2 instance;
+  if (list.contains(instance)) print('someFunction6'); // OK
+}
+
+abstract class ClassBase {}
+
+abstract class Mixin {}
+
+class DerivedClass2 extends ClassBase with Mixin {}
+```
+
+**GOOD:**
+```
+void someFunction7() {
+  List<Mixin> list = <Mixin>[];
+  DerivedClass3 instance;
+  if (list.contains(instance)) print('someFunction7'); // OK
+}
+
+abstract class ClassBase {}
+
+abstract class Mixin {}
+
+class DerivedClass3 extends ClassBase implements Mixin {}
+```
+''';
+
+class IterableContainsUnrelatedType extends LintRule {
+  _Visitor _visitor;
+
+  IterableContainsUnrelatedType() : super(
+          name: 'iterable_contains_unrelated_type',
+            description: _desc,
+            details: _details,
+      group: Group.errors,
+      maturity: Maturity.experimental) {
+    _visitor = new _Visitor(this);
+  }
+
+  @override
+  AstVisitor getVisitor() => _visitor;
+}
+
+class _Visitor extends SimpleAstVisitor {
+  final LintRule rule;
+  _Visitor(this.rule);
+
+  @override
+  void visitMethodInvocation(MethodInvocation node) {
+    if (!_isParameterizedContainsInvocation(node)) {
+      return;
+    }
+
+    ParameterizedType type = node.target.bestType;
+    Expression argument = node.argumentList.arguments.first;
+    if (unrelatedTypes(argument.bestType, _findIterableTypeArgument(type))) {
+      rule.reportLint(node);
+    }
+  }
+}
+
+bool _isParameterizedContainsInvocation(MethodInvocation node) =>
```

Comment:
Two problems:
1. The two `is` checks are redundant: every interface type is a parameterized type. I suspect you wanted the `is ParameterizedType` test to tell you whether the class _actually_ has type parameters rather than whether it is a kind of type that _could_ have type parameters, but this test won't do that. Instead, try `node.target.bestType.typeParameters.isNotEmpty`.
2. As noted before, this won't catch uses of contains in classes that implement `Iterable` when the target is implicitly `this`. I think that's fairly minor, but ought to be captured in a TODO.